<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Pierre Osteil">
<meta name="dcterms.date" content="2023-10-15">

<title>TP N3S6 - Introduction Génomique Fonctionnelle 2024</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="IGF_v1_files/libs/clipboard/clipboard.min.js"></script>
<script src="IGF_v1_files/libs/quarto-html/quarto.js"></script>
<script src="IGF_v1_files/libs/quarto-html/popper.min.js"></script>
<script src="IGF_v1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="IGF_v1_files/libs/quarto-html/anchor.min.js"></script>
<link href="IGF_v1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="IGF_v1_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="IGF_v1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="IGF_v1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="IGF_v1_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">TP N3S6 - Introduction Génomique Fonctionnelle 2024</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Pierre Osteil </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 15, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="table-des-matières" class="level1">
<h1>Table des Matières</h1>
<ol type="1">
<li><a href="#Intro">Introduction</a></li>
<li><a href="#Partie2">Analyse de données RT-qPCR avec R-Studio</a>
<ol type="1">
<li><a href="#Partie2.1">Charger les packages</a></li>
<li><a href="#Partie2.2">Charger les données</a></li>
<li><a href="#Partie2.3">Permutation de table</a></li>
<li><a href="#Partie2.4">Ajout de metadata</a></li>
<li><a href="#Partie2.5">Calcul des delta Ct</a></li>
<li><a href="#Partie2.6">Calcul des delta delta Ct</a></li>
<li><a href="#Partie2.7">Plot des résultats</a></li>
<li><a href="#Partie2.8">Annotation des groupes pour chaque echantillon</a></li>
</ol></li>
<li><a href="#Partie3">RNA-seq</a></li>
</ol>
</section>
<section id="introduction" class="level1">
<h1>Introduction <a name="Intro" href=""></a></h1>
<p>Lors de ce TP vous avez reçu un numéro. Ce numéro correspond à un type cellulaire chez un mammifère. Le jeu consiste, grâce aux données que vous avez en main, à identifier à quel type cellulaire votre numéro correspond. Pour cela nous allons effectuer un ensemble de lignes de commandes du logiciel R. Elle seront appliquées sur R Studio et l’ensemble des données produites pourra être visualisé grâce au programme Quarto, utilisé pour générer ce document. Afin d’améliorer vos compétences, vous pouvez consulter cet ouvrage: <a href="https://r4ds.had.co.nz/">R for Data Science</a></p>
<p>Le TP sera réalisé sur 8 séances de 1H30. Le compte rendu sera réalisé en remplissant les sections indiquées sur ce document et nous le corrigerons en TD.</p>
</section>
<section id="analyse-de-données-rt-qpcr-avec-r-studio" class="level1">
<h1>Analyse de données RT-qPCR avec R-Studio <a name="Partie2" href=""></a></h1>
<section id="charger-les-packages" class="level2">
<h2 class="anchored" data-anchor-id="charger-les-packages">1. Charger les packages <a name="Partie2.2" href=""></a></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Installer les packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("tidyverse")</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("ggpubr")</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'BiocManager' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version '3.17' is out-of-date; the current release version '3.18'
  is available with R version '4.3'; see https://bioconductor.org/install</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">BioC_mirror =</span> <span class="st">"http://bioconductor.org"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DESeq2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16 ucrt)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'DESeq2'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"apeglm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16 ucrt)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'apeglm'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"edgeR"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16 ucrt)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'edgeR'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"ComplexHeatmap"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16 ucrt)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'ComplexHeatmap'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"goseq"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16 ucrt)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'goseq'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"org.Mm.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16 ucrt)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'org.Mm.eg.db'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"clusterProfiler"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16 ucrt)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'clusterProfiler'</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#charger les package</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ggplot2' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'dplyr' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'stringr' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'lubridate' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.4.4     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors
Loading required package: stats4
Loading required package: BiocGenerics

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:lubridate':

    intersect, setdiff, union

The following objects are masked from 'package:dplyr':

    combine, intersect, setdiff, union

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min


Attaching package: 'S4Vectors'

The following objects are masked from 'package:lubridate':

    second, second&lt;-

The following objects are masked from 'package:dplyr':

    first, rename

The following object is masked from 'package:tidyr':

    expand

The following object is masked from 'package:utils':

    findMatches

The following objects are masked from 'package:base':

    expand.grid, I, unname

Loading required package: IRanges

Attaching package: 'IRanges'

The following object is masked from 'package:lubridate':

    %within%

The following objects are masked from 'package:dplyr':

    collapse, desc, slice

The following object is masked from 'package:purrr':

    reduce

The following object is masked from 'package:grDevices':

    windows

Loading required package: GenomicRanges
Loading required package: GenomeInfoDb
Loading required package: SummarizedExperiment
Loading required package: MatrixGenerics
Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'matrixStats' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'matrixStats'

The following object is masked from 'package:dplyr':

    count


Attaching package: 'MatrixGenerics'

The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars

Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.


Attaching package: 'Biobase'

The following object is masked from 'package:MatrixGenerics':

    rowMedians

The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(apeglm)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'edgeR' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: limma

Attaching package: 'limma'

The following object is masked from 'package:DESeq2':

    plotMA

The following object is masked from 'package:BiocGenerics':

    plotMA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: grid
========================================
ComplexHeatmap version 2.16.0
Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/
Github page: https://github.com/jokergoo/ComplexHeatmap
Documentation: http://jokergoo.github.io/ComplexHeatmap-reference

If you use it in published research, please cite either one:
- Gu, Z. Complex Heatmap Visualization. iMeta 2022.
- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional 
    genomic data. Bioinformatics 2016.


The new InteractiveComplexHeatmap package can directly export static 
complex heatmaps into an interactive Shiny app with zero effort. Have a try!

This message can be suppressed by:
  suppressPackageStartupMessages(library(ComplexHeatmap))
========================================</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(goseq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiasedUrn
Loading required package: geneLenDataBase

Attaching package: 'geneLenDataBase'

The following object is masked from 'package:S4Vectors':

    unfactor</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 methods overwritten by 'treeio':
  method              from    
  MRCA.phylo          tidytree
  MRCA.treedata       tidytree
  Nnode.treedata      tidytree
  Ntip.treedata       tidytree
  ancestor.phylo      tidytree
  ancestor.treedata   tidytree
  child.phylo         tidytree
  child.treedata      tidytree
  full_join.phylo     tidytree
  full_join.treedata  tidytree
  groupClade.phylo    tidytree
  groupClade.treedata tidytree
  groupOTU.phylo      tidytree
  groupOTU.treedata   tidytree
  inner_join.phylo    tidytree
  inner_join.treedata tidytree
  is.rooted.treedata  tidytree
  nodeid.phylo        tidytree
  nodeid.treedata     tidytree
  nodelab.phylo       tidytree
  nodelab.treedata    tidytree
  offspring.phylo     tidytree
  offspring.treedata  tidytree
  parent.phylo        tidytree
  parent.treedata     tidytree
  root.treedata       tidytree
  rootnode.phylo      tidytree
  sibling.phylo       tidytree
clusterProfiler v4.8.3  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/

If you use clusterProfiler in published research, please cite:
T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141

Attaching package: 'clusterProfiler'

The following object is masked from 'package:IRanges':

    slice

The following object is masked from 'package:S4Vectors':

    rename

The following object is masked from 'package:purrr':

    simplify

The following object is masked from 'package:stats':

    filter</code></pre>
</div>
</div>
</section>
<section id="charger-les-données" class="level2">
<h2 class="anchored" data-anchor-id="charger-les-données">2. Charger les données <a name="Partie2.2" href=""></a></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>( <span class="st">"C:/Users/piosteil/Desktop/Work/4-All_git/Cours/TP_Transcripto_N3S6/RT-qPCR/RT-qPCR_data.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">dec =</span> <span class="st">","</span>) <span class="co"># argument sep = "\t" pour tabulated, et dec = "," pour décimal </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><mark>Quelle sont les dimensions du jeu de données chargé? </mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(RTqPCR_dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 60 13</code></pre>
</div>
</div>
<p><mark>Combien de réplicat par echantillons ont été effectué? </mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Gene Ech1_Rep1 Ech1_Rep2 Ech1_Rep3 Ech2_Rep1 Ech2_Rep2 Ech2_Rep3 Ech3_Rep1
1 Foxa2      18.2      18.9      19.0      20.9      21.3      21.8      22.7
2 Foxa2      17.7      18.4      18.9      20.8      21.4      22.4      22.4
3 Foxa2      17.5      18.1      19.1      20.7      21.2      23.0      22.9
4 Gata6      15.3      16.3      16.3      17.9      18.5      19.1      19.4
5 Gata6      14.9      16.1      16.7      17.8      18.9      19.9      19.4
6 Gata6      15.0      16.2      16.8      18.1      19.0      19.8      19.4
  Ech3_Rep2 Ech3_Rep3 CTR_Rep1 CTR_Rep2 CTR_Rep3
1      23.4      24.1     31.9     32.3     32.5
2      23.0      24.1     30.6     32.0     32.6
3      22.6      24.4     31.0     31.6     33.1
4      20.0      21.3     29.0     29.5     30.2
5      20.3      21.3     29.0     30.6     30.5
6      20.5      21.6     28.9     29.3     30.4</code></pre>
</div>
</div>
</section>
<section id="permutation-de-table" class="level2">
<h2 class="anchored" data-anchor-id="permutation-de-table">3. Permutation de table <a name="Partie2.3" href=""></a></h2>
<p>Les tableaux fourni par les machines d’analyses ne sont souvent pas adapté à la production de plots. L’utilisation de ggplot2 requiert une organisation spéciale avec pour chaque colonne un type de donnée. Nous allons donc réorganiser le tableau. Pour cela les fonctions du package tidyr vont nous permettre de manipuler le tableau sans aucun risque d’erreur de mauvaises manipulations de copier/coller</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_dat_long <span class="ot">&lt;-</span> RTqPCR_dat <span class="sc">|&gt;</span> <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">13</span>, <span class="at">names_to =</span> <span class="st">"condition"</span>, <span class="at">values_to =</span> <span class="st">"Ct"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><mark>Quelle sont les dimensions de cette nouvelle table? </mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(RTqPCR_dat_long)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 720   3</code></pre>
</div>
</div>
<p><mark>Que fait pivot_longer()? </mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_dat_long)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  Gene  condition    Ct
  &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;
1 Foxa2 Ech1_Rep1  18.2
2 Foxa2 Ech1_Rep2  18.9
3 Foxa2 Ech1_Rep3  19  
4 Foxa2 Ech2_Rep1  20.9
5 Foxa2 Ech2_Rep2  21.3
6 Foxa2 Ech2_Rep3  21.8</code></pre>
</div>
</div>
</section>
<section id="ajout-de-metadata" class="level2">
<h2 class="anchored" data-anchor-id="ajout-de-metadata">4. Ajout de metadata <a name="Partie2.4" href=""></a></h2>
<p>Les données sont généralement brutes. Afin de présenter des plots compréhensible, il faut ajouter des métadata, qui sont des informations supplémentaire pour chaque échantillons.</p>
<p>Ici nous allons créer une colonne pour le nom de chaque échantillon et le nom de chaque réplicats.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Création d'un tableau à deux colonnes</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>metadata1 <span class="ot">&lt;-</span> <span class="fu">str_split_fixed</span>(RTqPCR_dat_long<span class="sc">$</span>condition, <span class="st">"_"</span>, <span class="dv">2</span>)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(metadata1) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"echantillon"</span>, <span class="st">"bioRep"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><mark>Que fait str_split_fixed()?</mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metadata1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     echantillon bioRep
[1,] "Ech1"      "Rep1"
[2,] "Ech1"      "Rep2"
[3,] "Ech1"      "Rep3"
[4,] "Ech2"      "Rep1"
[5,] "Ech2"      "Rep2"
[6,] "Ech2"      "Rep3"</code></pre>
</div>
</div>
<p><mark>Quelle sont les dimensions de cette nouvelle table? Est-ce attendu? Commentez </mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(metadata1) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 720   2</code></pre>
</div>
</div>
<p>Combinez avec le tableau de résultats</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_dat_long_meta <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(RTqPCR_dat_long, metadata1)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_dat_long_meta)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Gene condition   Ct echantillon bioRep
1 Foxa2 Ech1_Rep1 18.2        Ech1   Rep1
2 Foxa2 Ech1_Rep2 18.9        Ech1   Rep2
3 Foxa2 Ech1_Rep3 19.0        Ech1   Rep3
4 Foxa2 Ech2_Rep1 20.9        Ech2   Rep1
5 Foxa2 Ech2_Rep2 21.3        Ech2   Rep2
6 Foxa2 Ech2_Rep3 21.8        Ech2   Rep3</code></pre>
</div>
</div>
<p>Nous avons également une autre information qui va s’avérer utile pour la suite de notr analyse, qui est une ontologie sur les trois feuillet embryonnaires. Cette information se trouve dans un autre fichier “fonction_gène.txt”. Nous allons l’ouvrir et incorporer ces données aux tableaux de résultats</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>metadata2 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"C:/Users/piosteil/Desktop/Work/4-All_git/Cours/TP_Transcripto_N3S6/RT-qPCR/fonction_gène.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><mark>Quelle sont les dimensions de cette nouvelle table? Est-ce attendu? Commentez</mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(metadata2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20  2</code></pre>
</div>
</div>
<p>Afin de compenser le nombre de ligne nous allons utiliser une autre fonction, merge()</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_dat_long_meta2 <span class="ot">&lt;-</span> <span class="fu">merge</span>(RTqPCR_dat_long_meta, metadata2, <span class="at">by =</span> <span class="st">"Gene"</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(RTqPCR_dat_long_meta2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Gene condition   Ct echantillon bioRep     Fonction
715 Zfp42 Ech3_Rep1 31.8        Ech3   Rep1 Pluripotence
716 Zfp42 Ech3_Rep2 31.7        Ech3   Rep2 Pluripotence
717 Zfp42 Ech3_Rep3 33.0        Ech3   Rep3 Pluripotence
718 Zfp42  CTR_Rep1 15.0         CTR   Rep1 Pluripotence
719 Zfp42  CTR_Rep2 15.4         CTR   Rep2 Pluripotence
720 Zfp42  CTR_Rep3 16.7         CTR   Rep3 Pluripotence</code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(RTqPCR_dat_long_meta2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 720   6</code></pre>
</div>
</div>
<p><mark>Comment fonctionne la fonction merge()?</mark></p>
</section>
<section id="calcul-des-delta-ct" class="level2">
<h2 class="anchored" data-anchor-id="calcul-des-delta-ct">5. Calcul des delta Ct <a name="Partie2.5" href=""></a></h2>
<p>La première étape va être de calculer les medianes de chaques réplicats techniques. Pour cela la table va être réduite.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_median <span class="ot">&lt;-</span> RTqPCR_dat_long_meta2 <span class="sc">|&gt;</span> <span class="fu">group_by</span>(condition, Gene) <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">MedCt =</span> <span class="fu">median</span>(Ct))</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_median)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
# Groups:   condition, Gene [6]
  Gene  condition    Ct echantillon bioRep Fonction  MedCt
  &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;     &lt;dbl&gt;
1 Foxa2 Ech1_Rep1  18.2 Ech1        Rep1   Endoderme  17.7
2 Foxa2 Ech1_Rep2  18.9 Ech1        Rep2   Endoderme  18.4
3 Foxa2 Ech1_Rep3  19   Ech1        Rep3   Endoderme  19  
4 Foxa2 Ech2_Rep1  20.9 Ech2        Rep1   Endoderme  20.8
5 Foxa2 Ech2_Rep2  21.3 Ech2        Rep2   Endoderme  21.3
6 Foxa2 Ech2_Rep3  21.8 Ech2        Rep3   Endoderme  22.4</code></pre>
</div>
</div>
<p><mark>Pourquoi calcules-t-on la médiane et non pas la moyenne pour les réplicats techniques? </mark></p>
<p>Maintenant que les réplicats techniques ont été utilisés, on peut les enlever</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_median_clean <span class="ot">&lt;-</span> RTqPCR_median <span class="sc">|&gt;</span> <span class="fu">group_by</span>(condition, Gene) <span class="sc">|&gt;</span> <span class="fu">distinct</span>(condition, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="co">#on va selectioner l'échantillon median à chaque fois</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(RTqPCR_median_clean)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
# Groups:   condition, Gene [6]
  Gene  condition    Ct echantillon bioRep Fonction     MedCt
  &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;        &lt;dbl&gt;
1 Zfp42 Ech3_Rep1  31.7 Ech3        Rep1   Pluripotence  31.7
2 Zfp42 Ech3_Rep2  32.3 Ech3        Rep2   Pluripotence  32.1
3 Zfp42 Ech3_Rep3  32.8 Ech3        Rep3   Pluripotence  32.8
4 Zfp42 CTR_Rep1   15   CTR         Rep1   Pluripotence  15  
5 Zfp42 CTR_Rep2   16   CTR         Rep2   Pluripotence  15.6
6 Zfp42 CTR_Rep3   16.6 CTR         Rep3   Pluripotence  16.6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(RTqPCR_median_clean)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 240   7</code></pre>
</div>
</div>
<p><mark> A quoi sert la fonction “distinct”? </mark></p>
<p>Nous allons extraire les données du gènes de ménage (GDM), et calculer la moyenne pour chaque échantillon</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_median_GDM <span class="ot">&lt;-</span> RTqPCR_median <span class="sc">|&gt;</span> <span class="fu">filter</span>(Fonction <span class="sc">==</span> <span class="st">"GDM"</span>) </span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_median_GDM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
# Groups:   condition, Gene [6]
  Gene  condition    Ct echantillon bioRep Fonction MedCt
  &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;
1 TBP   Ech1_Rep1  23   Ech1        Rep1   GDM       22.9
2 TBP   Ech1_Rep2  23.9 Ech1        Rep2   GDM       23.6
3 TBP   Ech1_Rep3  23.5 Ech1        Rep3   GDM       24  
4 TBP   Ech2_Rep1  23   Ech2        Rep1   GDM       22.9
5 TBP   Ech2_Rep2  22.6 Ech2        Rep2   GDM       23  
6 TBP   Ech2_Rep3  24   Ech2        Rep3   GDM       24  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(RTqPCR_median_GDM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 36</code></pre>
</div>
</div>
<p>Nous allons maintenant produire notre premier plot. Il s’agit de représenter la distibution des valeur moyenne de Ct en fonction des réplicats biologiques et des différent échantillons. Pour ceci nous allons utiliser un boxplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(RTqPCR_median_GDM, <span class="fu">aes</span>(<span class="at">x =</span> echantillon, <span class="at">y =</span> MedCt))<span class="sc">+</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">fill=</span><span class="st">"slateblue"</span>, <span class="at">alpha=</span><span class="fl">0.2</span>)<span class="sc">+</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">40</span>))<span class="sc">+</span> </span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_jitter</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><mark> Quelle conclusion tirez-vous de ce plot? </mark></p>
<p>L’objectif va être de soustraire les valeurs de Ct obtenu sur le GDM au valeurs de nos gènes. Dans un premier temps nous ajoutons la colonne MeanCT_GDM à notre tableau principal</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co">#on ne garde que les deux colonne d'interêt</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>RTqPCR_median_GDM_clean <span class="ot">&lt;-</span> RTqPCR_median_GDM <span class="sc">|&gt;</span> <span class="fu">select</span>(Gene, echantillon, condition, MedCt) <span class="sc">|&gt;</span> <span class="fu">distinct</span>(condition, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>RTqPCR_mean_GDM <span class="ot">&lt;-</span> RTqPCR_median_GDM_clean <span class="sc">|&gt;</span> <span class="fu">group_by</span>(echantillon) <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">MeanCt_GDM =</span> <span class="fu">mean</span>(MedCt)) <span class="sc">|&gt;</span> <span class="fu">distinct</span>(echantillon, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> <span class="fu">ungroup</span>()<span class="sc">|&gt;</span> <span class="fu">select</span>(echantillon, MeanCt_GDM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><mark> Discutez des deux opérations ci dessus. Quel est le but de la deuxième opération? </mark></p>
<p>on fucionne les deux tables pour ajouter les moyenne du GDM</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_dCt <span class="ot">&lt;-</span> <span class="fu">merge</span>(RTqPCR_median_clean, RTqPCR_mean_GDM,</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"echantillon"</span>))</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_dCt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  echantillon  Gene condition   Ct bioRep  Fonction MedCt MeanCt_GDM
1         CTR Mesp1  CTR_Rep1 31.7   Rep1 Mesoderme  31.5       23.3
2         CTR Mesp1  CTR_Rep2 32.7   Rep2 Mesoderme  32.5       23.3
3         CTR Hnf4a  CTR_Rep1 33.2   Rep1 Endoderme  32.9       23.3
4         CTR Hnf4a  CTR_Rep2 34.3   Rep2 Endoderme  34.0       23.3
5         CTR Foxa2  CTR_Rep1 31.9   Rep1 Endoderme  31.0       23.3
6         CTR Foxa2  CTR_Rep2 32.3   Rep2 Endoderme  32.0       23.3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(RTqPCR_dCt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 240   8</code></pre>
</div>
</div>
<p>On peut maintenant soustraire les MeanCT_GDM le MedCt de nos replicats techniques pour chaques gène à la moyenne de la valeur du GDM pour cet échantillon</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_dCt <span class="ot">&lt;-</span> RTqPCR_dCt <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">dCt =</span> MedCt<span class="sc">-</span>MeanCt_GDM)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_dCt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  echantillon  Gene condition   Ct bioRep  Fonction MedCt MeanCt_GDM  dCt
1         CTR Mesp1  CTR_Rep1 31.7   Rep1 Mesoderme  31.5       23.3  8.2
2         CTR Mesp1  CTR_Rep2 32.7   Rep2 Mesoderme  32.5       23.3  9.2
3         CTR Hnf4a  CTR_Rep1 33.2   Rep1 Endoderme  32.9       23.3  9.6
4         CTR Hnf4a  CTR_Rep2 34.3   Rep2 Endoderme  34.0       23.3 10.7
5         CTR Foxa2  CTR_Rep1 31.9   Rep1 Endoderme  31.0       23.3  7.7
6         CTR Foxa2  CTR_Rep2 32.3   Rep2 Endoderme  32.0       23.3  8.7</code></pre>
</div>
</div>
<p>Faisons un peu de nettoyage</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(metadata1, metadata2, RTqPCR_dat, RTqPCR_dat_long, RTqPCR_dat_long_meta, RTqPCR_dat_long_meta2, RTqPCR_mean_GDM, RTqPCR_median,  RTqPCR_median_clean, RTqPCR_median_GDM, RTqPCR_median_GDM_clean) <span class="co">#on ne garde que la dernière table</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="calcul-des-delta-delta-ct" class="level2">
<h2 class="anchored" data-anchor-id="calcul-des-delta-delta-ct">6. Calcul des delta delta Ct<a name="Partie2.6" href=""></a></h2>
<p>Nous avons normalisé les données sur le GDM, c’est à dire sur la quantité de cDNA (ARN RT) chargé dans la plaque. Maintenant les valeurs de dCt ne sont pas facile à comparer. Pour cela nous allons normaliser une deuxième fois (le deuxième “delta”), cette fois-ci pas sur un gène mais sur un échantillon, l’échantillon Contrôle (CTR).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extraction des données de l'échantillon contrôle et calcul de la moyenne</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>CTR_data <span class="ot">&lt;-</span> RTqPCR_dCt <span class="sc">|&gt;</span> <span class="fu">filter</span>(echantillon <span class="sc">==</span> <span class="st">"CTR"</span>) <span class="sc">|&gt;</span> <span class="fu">group_by</span>(Gene)<span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">MeanCt_CTR =</span> <span class="fu">mean</span>(MedCt)) <span class="sc">|&gt;</span> <span class="fu">distinct</span>(Gene, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> <span class="fu">ungroup</span>() <span class="sc">|&gt;</span> <span class="fu">select</span>(Gene, MeanCt_CTR, MeanCt_GDM)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(CTR_data)[<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="st">"MeanCt_CTR_GDM"</span></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(CTR_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  Gene  MeanCt_CTR MeanCt_CTR_GDM
  &lt;chr&gt;      &lt;dbl&gt;          &lt;dbl&gt;
1 Mesp1       32.3           23.3
2 Hnf4a       33.7           23.3
3 Foxa2       31.9           23.3
4 Mixl1       32.3           23.3
5 Klf4        15.6           23.3
6 Gata6       29.6           23.3</code></pre>
</div>
</div>
<p><mark> Discutez de l’opération ci dessus. </mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>RTqPCR_ddCt <span class="ot">&lt;-</span> <span class="fu">merge</span>(RTqPCR_dCt, CTR_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"Gene"</span>))</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>RTqPCR_ddCt <span class="ot">&lt;-</span> RTqPCR_ddCt <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">TWO_ddCt =</span> <span class="dv">2</span><span class="sc">^-</span>(dCt <span class="sc">-</span> (MeanCt_CTR <span class="sc">-</span> MeanCt_CTR_GDM)))</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_ddCt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Gene echantillon condition   Ct bioRep  Fonction MedCt MeanCt_GDM        dCt
1 Foxa2        Ech3 Ech3_Rep3 24.1   Rep3 Endoderme  24.1   23.36667  0.7333333
2 Foxa2         CTR  CTR_Rep1 31.9   Rep1 Endoderme  31.0   23.30000  7.7000000
3 Foxa2        Ech3 Ech3_Rep2 23.4   Rep2 Endoderme  23.0   23.36667 -0.3666667
4 Foxa2        Ech2 Ech2_Rep2 21.3   Rep2 Endoderme  21.3   23.30000 -2.0000000
5 Foxa2         CTR  CTR_Rep2 32.3   Rep2 Endoderme  32.0   23.30000  8.7000000
6 Foxa2        Ech2 Ech2_Rep1 20.9   Rep1 Endoderme  20.8   23.30000 -2.5000000
  MeanCt_CTR MeanCt_CTR_GDM     TWO_ddCt
1   31.86667           23.3  228.0700718
2   31.86667           23.3    1.8234450
3   31.86667           23.3  488.8789012
4   31.86667           23.3 1516.6438935
5   31.86667           23.3    0.9117225
6   31.86667           23.3 2144.8583635</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(RTqPCR_ddCt, <span class="st">"C:/Users/piosteil/Desktop/Work/4-All_git/Cours/TP_Transcripto_N3S6/RT-qPCR/RTqPCR_ddCt.csv"</span>)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(CTR_data, RTqPCR_dCt) <span class="co"># on nettoie</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-des-résultats" class="level2">
<h2 class="anchored" data-anchor-id="plot-des-résultats">7. Plot des résultats <a name="Partie2.7" href=""></a></h2>
<p>Dans cette partie on va pouvoir générer des plots pour enfin connaître d’où provienne nos échantillons. Les données représentent 240 observations (12 échantillons par 20 gènes), nous allons pouvoir réalisé une ACP.</p>
<p><mark> Qu’est-ce qu’une ACP ? </mark></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co">#on doit pivoter le tableau pour la pca</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="co">#avant cela enlevons les colonnes inutiles</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>RTqPCR_dCt_only <span class="ot">&lt;-</span> RTqPCR_ddCt <span class="sc">|&gt;</span> <span class="fu">select</span>(condition, Gene, dCt)</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a><span class="co"># on utilise pivot_wider cette fois ci</span></span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>RTqPCR_dCt_wide <span class="ot">&lt;-</span> RTqPCR_dCt_only <span class="sc">|&gt;</span> <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> Gene, <span class="at">values_from =</span> dCt)</span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a><span class="co">#write.csv(RTqPCR_dCt_wide, "./TP_Transcripto_N3S6/RT-qPCR/RTqPCR_dCt_wider.csv")</span></span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_dCt_wide)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 21
  condition  Foxa2 Gata6  Hand2  Hnf4a  Klf4  Krt14  Mesp1  Mixl1 Nanog Nestin
  &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 Ech3_Rep3  0.733 -2.07  0.633  2.33   9.63 -7.07   1.03   1.03   9.63 -7.37 
2 CTR_Rep1   7.7    5.7   8      9.6   -8.4   8.5    8.2    8.2   -8.3   8.5  
3 Ech3_Rep2 -0.367 -3.07 -0.167  1.63   8.83 -7.87   0.333  0.333  8.83 -7.77 
4 Ech2_Rep2 -2     -4.40 -7.6    0.300  8.9   0     -8     -8.1    8.9  -0.200
5 CTR_Rep2   8.7    6.2   8.7   10.7   -7.7   9.3    9.2    9.1   -7.4   9.1  
6 Ech2_Rep1 -2.5   -5.4  -8.3   -0.300  8    -0.900 -8.7   -8.6    8.2  -0.600
# ℹ 10 more variables: Neurog1 &lt;dbl&gt;, Pax6 &lt;dbl&gt;, Pdx1 &lt;dbl&gt;, Pou5F1 &lt;dbl&gt;,
#   Sox17 &lt;dbl&gt;, Sox2 &lt;dbl&gt;, TBP &lt;dbl&gt;, TBra &lt;dbl&gt;, Tbx6 &lt;dbl&gt;, Zfp42 &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>pca_rtqpcr <span class="ot">&lt;-</span> RTqPCR_dCt_wide <span class="sc">|&gt;</span> <span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">|&gt;</span> <span class="fu">prcomp</span>()</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>RTqPCR_dCt_ACP <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(RTqPCR_dCt_wide, pca_rtqpcr<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_dCt_ACP)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  condition      Foxa2     Gata6      Hand2     Hnf4a      Klf4     Krt14
1 Ech3_Rep3  0.7333333 -2.066667  0.6333333  2.333333  9.633333 -7.066667
2  CTR_Rep1  7.7000000  5.700000  8.0000000  9.600000 -8.400000  8.500000
3 Ech3_Rep2 -0.3666667 -3.066667 -0.1666667  1.633333  8.833333 -7.866667
4 Ech2_Rep2 -2.0000000 -4.400000 -7.6000000  0.300000  8.900000  0.000000
5  CTR_Rep2  8.7000000  6.200000  8.7000000 10.700000 -7.700000  9.300000
6 Ech2_Rep1 -2.5000000 -5.400000 -8.3000000 -0.300000  8.000000 -0.900000
       Mesp1      Mixl1     Nanog    Nestin   Neurog1      Pax6       Pdx1
1  1.0333333  1.0333333  9.633333 -7.366667 -7.166667 -6.866667  0.7333333
2  8.2000000  8.2000000 -8.300000  8.500000  6.000000  8.200000  8.2000000
3  0.3333333  0.3333333  8.833333 -7.766667 -8.166667 -7.766667 -0.3666667
4 -8.0000000 -8.1000000  8.900000 -0.200000  0.400000  0.710000 -1.7000000
5  9.2000000  9.1000000 -7.400000  9.100000  7.100000  8.800000  8.8000000
6 -8.7000000 -8.6000000  8.200000 -0.600000 -0.300000  0.010000 -2.3000000
     Pou5F1      Sox17      Sox2        TBP       TBra       Tbx6     Zfp42
1  9.633333  0.8333333 -6.866667  0.7333333  1.0333333  0.8333333  9.433333
2 -7.900000  5.7000000 -7.400000 -0.7000000  8.3000000  8.3000000 -8.300000
3  8.733333  0.1333333 -7.766667 -0.1666667  0.2333333  0.3333333  8.733333
4  8.800000 -0.2000000 -0.100000 -0.3000000 -7.7000000 -7.8000000  8.700000
5 -7.400000  6.3000000 -6.600000 -0.1000000  8.8000000  9.1000000 -7.700000
6  8.200000 -1.1000000 -2.200000 -0.4000000 -8.3000000 -8.4000000  7.900000
        PC1         PC2       PC3
1  10.73089 -13.2482946 4.0779416
2 -38.47091   0.8796006 0.9517404
3  12.04431 -13.3692460 4.1727334
4  16.28734  11.8140398 5.7514030
5 -39.86492   0.9073582 0.5352716
6  17.20867  11.3653818 6.1388310</code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(RTqPCR_dCt_ACP, <span class="fu">aes</span>(PC1, PC2, <span class="at">color =</span> condition)) <span class="sc">+</span> </span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><mark> Decrivez le plot obtenu </mark></p>
<p><mark> Realisez un plot avec votre echantillon en couleur et le reste en gris </mark></p>
<p>Pour ameliorer vos plots: <a href="https://www.maths.usyd.edu.au/u/UG/SM/STAT3022/r/current/Misc/data-visualization-2.1.pdf">cheatsheet</a></p>
</section>
<section id="annotation-des-groupes-pour-chaque-echantillon" class="level2">
<h2 class="anchored" data-anchor-id="annotation-des-groupes-pour-chaque-echantillon">8. Annotation des groupes pour chaque echantillon <a name="Partie2.8" href=""></a></h2>
<section id="calul-des-p.values" class="level3">
<h3 class="anchored" data-anchor-id="calul-des-p.values">8.1 calul des p.values</h3>
<p>Avant de plotter l’expression de chaque gene, on va calculer la p.value. Ensuite on pourra plotter seulement les genes significativement, differentiellement exprimés.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>my_sample <span class="ot">&lt;-</span> <span class="st">"Ech2"</span> </span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>RTqPCR_ddCt_Ech <span class="ot">&lt;-</span> RTqPCR_ddCt <span class="sc">|&gt;</span> <span class="fu">filter</span>(echantillon <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"CTR"</span>, my_sample)) <span class="sc">|&gt;</span> <span class="fu">group_by</span>(Gene) <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">pval =</span> <span class="fu">list</span>(<span class="fu">t.test</span>(dCt[echantillon <span class="sc">==</span> my_sample], dCt[echantillon <span class="sc">==</span> <span class="st">"CTR"</span>])<span class="sc">$</span>p.value))</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RTqPCR_ddCt_Ech)<span class="sc">$</span>pval</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] 9.848311e-05

[[2]]
[1] 9.848311e-05

[[3]]
[1] 9.848311e-05

[[4]]
[1] 9.848311e-05

[[5]]
[1] 9.848311e-05

[[6]]
[1] 9.848311e-05</code></pre>
</div>
</div>
</section>
<section id="plot-des-genes" class="level3">
<h3 class="anchored" data-anchor-id="plot-des-genes">8.2 Plot des genes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(RTqPCR_ddCt_Ech, <span class="fu">aes</span>(<span class="at">x=</span>echantillon, <span class="at">y =</span> TWO_ddCt, <span class="at">fill =</span> Fonction))<span class="sc">+</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>()<span class="sc">+</span> </span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="st">"Gene"</span>)<span class="sc">+</span></span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Cyan"</span>, <span class="st">"Green"</span>, <span class="st">"black"</span>,<span class="st">"Red"</span>, <span class="st">"Magenta"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
</section>
<section id="rna-seq" class="level1">
<h1>3. RNA-seq <a name="Partie3" href=""></a></h1>
<section id="charger-les-donnees" class="level2">
<h2 class="anchored" data-anchor-id="charger-les-donnees">Charger les donnees <a name="Partie3.1" href=""></a></h2>
<p>Maintenant que nous avons identifié le tissus germinale nous allons pouvoir constituer un groupe d’étude pour identifier le tissus final. Pour cela, nous avons a disposition les résultats de RNA-seq sur deux répliquats de chaque échantillon. Ces résultats sont présenté de la manière suivante: 1 colonne par répliquat, et une ligne par gène. Vous allez selectionner deux échantillons: le votre et un autre aléatoire. Comme chaque échantillon possède deux répliquats, vous aller travailler avec un tableau à 4 colonnes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co">#data</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>RNA_dat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">read.csv2</span>(<span class="st">"C:/Users/piosteil/Desktop/Work/4-All_git/Cours/TP_Transcripto_N3S6/RNA-seq/RNAseq_countTable.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">row.names=</span> <span class="dv">1</span>))</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RNA_dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Cerebellum Cerebellum.1 Kidney Kidney.1 Liver Liver.1
Gnai3          2           33      1      201   154     166
Pbsn           0            0      0        0     0       0
Cdc45          0            0      0        0     0       2
H19            0            1      0        1     0       0
Scml2          0            1      0        1     0       0
Apoh           0            0      0       13  1497    2442</code></pre>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co">#metadata </span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>sampleCondition <span class="ot">&lt;-</span> <span class="fu">read.csv2</span>(<span class="st">"C:/Users/piosteil/Desktop/Work/4-All_git/Cours/TP_Transcripto_N3S6/RNA-seq/SRAs/GSE30352_All_SRAs.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>my_sample <span class="ot">&lt;-</span> <span class="st">"Liver"</span></span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>ref_sample <span class="ot">&lt;-</span> <span class="st">"Kidney"</span> <span class="co">#un echantillon aléatoire qui n'est pas le votre</span></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a><span class="co">#other_samples &lt;- sampleCondition  |&gt; select(source_name) |&gt; filter(!(source_name %in% c(my_sample, ref_sample)))</span></span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a><span class="co">#other_samples &lt;- unique(c(other_sample$source_name))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(RNA_dat, <span class="fu">DataFrame</span>(sampleCondition), <span class="sc">~</span> source_name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in
design formula are characters, converting to factors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 55109 6 
metadata(1): version
assays(1): counts
rownames(55109): Gnai3 Pbsn ... CT485612.1 AC158307.1
rowData names(0):
colnames(6): Cerebellum Cerebellum.1 ... Liver Liver.1
colData names(33): Run Age ... SRA.Study tissue_source</code></pre>
</div>
</div>
<p>Inspecte le dataset. Combien d’echantillon, de genes et metadata?</p>
</section>
<section id="nettoyage-echantillon-avec-peu-de-count" class="level2">
<h2 class="anchored" data-anchor-id="nettoyage-echantillon-avec-peu-de-count">Nettoyage echantillon avec peu de count</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>smallestGroupSize <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(dds) <span class="sc">&gt;=</span> <span class="dv">10</span>) <span class="sc">&gt;=</span> smallestGroupSize</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> dds[keep,]</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 5364 6 
metadata(1): version
assays(1): counts
rownames(5364): Gnai3 Apoh ... Gm10503 9130016M20Rik
rowData names(0):
colnames(6): Cerebellum Cerebellum.1 ... Liver Liver.1
colData names(33): Run Age ... SRA.Study tissue_source</code></pre>
</div>
</div>
<p>Inspecte le dataset. Combien de genes ont ete filtre?</p>
</section>
<section id="genes-differentiellement-exprimes" class="level2">
<h2 class="anchored" data-anchor-id="genes-differentiellement-exprimes">Genes differentiellement exprimes</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>dds<span class="sc">$</span>source_name <span class="ot">&lt;-</span> <span class="fu">factor</span>(dds<span class="sc">$</span>source_name)</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">'source_name'</span>, my_sample, ref_sample))</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): source_name Liver vs Kidney 
Wald test p-value: source_name Liver vs Kidney 
DataFrame with 5364 rows and 6 columns
               baseMean log2FoldChange     lfcSE      stat      pvalue
              &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
Gnai3          31.84612       1.025296  0.967932  1.059264 2.89479e-01
Apoh          234.83128       8.290712  1.128039  7.349670 1.98697e-13
Narf           12.39308       1.580219  1.192694  1.324916 1.85199e-01
Klf6            6.39722      -1.350349  1.432011 -0.942974 3.45694e-01
Cox5a          10.44934       0.618385  1.225381  0.504647 6.13807e-01
...                 ...            ...       ...       ...         ...
Gm49336         5.95100       0.536179   1.45363  0.368855    0.712236
Pnp             3.99500       1.118263   1.48860  0.751220    0.452520
Riok2           4.15521       0.349275   1.45462  0.240115    0.810241
Gm10503         3.02743      -0.649835   1.94345 -0.334371    0.738099
9130016M20Rik  21.01543       1.809556   1.68803  1.071996    0.283722
                     padj
                &lt;numeric&gt;
Gnai3         9.26766e-01
Apoh          1.26967e-10
Narf          7.77698e-01
Klf6                   NA
Cox5a         9.94410e-01
...                   ...
Gm49336                NA
Pnp                    NA
Riok2                  NA
Gm10503                NA
9130016M20Rik    0.917966</code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Shrinkage of effect size (LFC estimates) is useful for visualization and ranking of genes. </span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>resLFC <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef=</span><span class="st">"source_name_Kidney_vs_Cerebellum"</span>, <span class="at">type=</span><span class="st">"apeglm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895</code></pre>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>resLFC</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MAP): source name Kidney vs Cerebellum 
Wald test p-value: source name Kidney vs Cerebellum 
DataFrame with 5364 rows and 5 columns
               baseMean log2FoldChange     lfcSE     pvalue      padj
              &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
Gnai3          31.84612      0.4289446  0.690818  0.2420037  0.633240
Apoh          234.83128      0.5473938  0.948278  0.0782913  0.382799
Narf           12.39308     -0.1801204  0.635356  0.5632992  0.843881
Klf6            6.39722      0.1781809  0.649700  0.5633148        NA
Cox5a          10.44934      0.0584684  0.615098  0.8461193  0.961007
...                 ...            ...       ...        ...       ...
Gm49336         5.95100     -0.1235323  0.645449 0.68526461        NA
Pnp             3.99500      0.3278576  0.756771 0.18161042        NA
Riok2           4.15521      0.0441583  0.643429 0.80348608        NA
Gm10503         3.02743     -0.0233358  0.657801 0.97991852        NA
9130016M20Rik  21.01543      0.5602375  1.045533 0.00752006 0.0885104</code></pre>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>resOrdered <span class="ot">&lt;-</span> res[<span class="fu">order</span>(res<span class="sc">$</span>log2FoldChange),]</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv2</span>(resOrdered, <span class="st">"C:/Users/piosteil/Desktop/Work/4-All_git/Cours/TP_Transcripto_N3S6/RNA-seq/DESeq_Kidney-Cerebellum.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 5364 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 153, 2.9%
LFC &lt; 0 (down)     : 31, 0.58%
outliers [1]       : 0, 0%
low counts [2]     : 2808, 52%
(mean count &lt; 10)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(res<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 141</code></pre>
</div>
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(resLFC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 5364 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 139, 2.6%
LFC &lt; 0 (down)     : 113, 2.1%
outliers [1]       : 0, 0%
low counts [2]     : 2704, 50%
(mean count &lt; 9)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(resLFC<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 163</code></pre>
</div>
</div>
</section>
<section id="explorer-les-datas" class="level2">
<h2 class="anchored" data-anchor-id="explorer-les-datas">Explorer les datas</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co">#with DESeq2 package</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>DESeq2<span class="sc">::</span><span class="fu">plotMA</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>DESeq2<span class="sc">::</span><span class="fu">plotMA</span>(resLFC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-34-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co"># with ggpubr package </span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggmaplot</span>(res,</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">fdr =</span> <span class="fl">0.05</span>,</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">fc =</span> <span class="dv">2</span>, <span class="at">size =</span> <span class="dv">2</span>,   <span class="at">top =</span> <span class="dv">50</span></span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-34-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggmaplot</span>(resLFC,</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">fdr =</span> <span class="fl">0.05</span>,</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">fc =</span> <span class="dv">2</span>, <span class="at">size =</span> <span class="dv">2</span>,   <span class="at">top =</span> <span class="dv">50</span></span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-34-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PCA </span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>vsd <span class="ot">&lt;-</span> <span class="fu">vst</span>(dds, <span class="at">blind =</span> <span class="cn">FALSE</span>)</span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(vsd, <span class="at">intgroup =</span> <span class="st">'source_name'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PCA with ggplot2</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>pcaData <span class="ot">&lt;-</span> <span class="fu">plotPCA</span>(vsd, <span class="at">intgroup=</span><span class="st">'source_name'</span>, <span class="at">returnData=</span><span class="cn">TRUE</span>)</span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>percentVar <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">attr</span>(pcaData, <span class="st">"percentVar"</span>))</span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggscatter</span>(pcaData, <span class="st">'PC1'</span>, <span class="st">'PC2'</span>, <span class="at">color=</span><span class="st">'source_name'</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC1: "</span>,percentVar[<span class="dv">1</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span></span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC2: "</span>,percentVar[<span class="dv">2</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span> </span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-35-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Heatmap</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>resDDS <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(res<span class="sc">$</span>log2FoldChange, res<span class="sc">$</span>padj)</span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(resDDS) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res)</span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>resSIG <span class="ot">&lt;-</span> resDDS <span class="sc">|&gt;</span> <span class="fu">filter</span>(res<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(resSIG)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 141   2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>resSIGOrdered <span class="ot">&lt;-</span> resSIG[<span class="fu">order</span>(<span class="fu">abs</span>(resSIG<span class="sc">$</span><span class="st">`</span><span class="at">res$log2FoldChange</span><span class="st">`</span>)),]</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>resSIGOrdered_top50 <span class="ot">&lt;-</span> <span class="fu">head</span>(resSIGOrdered, <span class="dv">200</span>)</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>RNA_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(RNA_dat)</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>RNA_df_SIG <span class="ot">&lt;-</span> RNA_df<span class="sc">|&gt;</span> <span class="fu">filter</span>(<span class="fu">row.names</span>(RNA_df) <span class="sc">%in%</span> <span class="fu">row.names</span>(resSIGOrdered_top50))</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a>hm <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(<span class="at">matrix =</span> <span class="fu">scale</span>(RNA_df_SIG),</span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">clustering_method_rows =</span> <span class="st">"centroid"</span>,</span>
<span id="cb139-11"><a href="#cb139-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>,</span>
<span id="cb139-12"><a href="#cb139-12" aria-hidden="true" tabindex="-1"></a>               <span class="at">clustering_method_columns =</span> <span class="st">"centroid"</span>,</span>
<span id="cb139-13"><a href="#cb139-13" aria-hidden="true" tabindex="-1"></a>               <span class="at">show_row_names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb139-14"><a href="#cb139-14" aria-hidden="true" tabindex="-1"></a>               <span class="at">show_column_names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb139-15"><a href="#cb139-15" aria-hidden="true" tabindex="-1"></a>hm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co">#GOseq from vignette</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="co">#test supported genome</span></span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>supported <span class="ot">&lt;-</span> <span class="fu">supportedOrganisms</span>()[<span class="fu">supportedOrganisms</span>()<span class="sc">$</span>Genome<span class="sc">==</span><span class="st">"mm9"</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: rtracklayer</code></pre>
</div>
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract differentially expressed genes from "res" of DESeq2 result</span></span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> res<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(res<span class="sc">$</span>padj) <span class="sc">&amp;</span> res<span class="sc">$</span>log2FoldChange <span class="sc">&lt;</span> <span class="dv">0</span> </span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(genes) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res)</span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Probability Weighted Function </span></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a>pwf<span class="ot">=</span><span class="fu">nullp</span>(genes,<span class="st">"mm9"</span>,<span class="st">"geneSymbol"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading mm9 length data...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in pcls(G): initial point very close to some inequality constraints</code></pre>
</div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pwf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      DEgenes bias.data         pwf
Gnai3   FALSE      3220 0.003954547
Apoh    FALSE      1194 0.002627870
Narf    FALSE      4423 0.004734978
Klf6    FALSE      4221 0.004602315
Cox5a   FALSE       697 0.002298959
Fer     FALSE        NA          NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Gene Ontology</span></span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>GO_wall<span class="ot">=</span><span class="fu">goseq</span>(pwf,<span class="st">"mm9"</span>,<span class="st">"geneSymbol"</span>, <span class="at">test.cats=</span><span class="st">"GO:BP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Fetching GO annotations...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: AnnotationDbi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'AnnotationDbi'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:clusterProfiler':

    select</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    select</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For 94 genes, we could not find any categories. These genes will be excluded.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>To force their use, please run with use_genes_without_cat=TRUE (see documentation).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>This was the default behavior for version 1.15.1 and earlier.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating the p-values...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(GO_wall, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        category over_represented_pvalue under_represented_pvalue numDEInCat
11250 GO:1904684            7.097327e-05                0.9999999          2
11386 GO:1905049            1.283461e-04                0.9999996          2
11249 GO:1904683            1.939761e-04                0.9999992          2
5547  GO:0043576            2.420431e-04                0.9999990          2
11385 GO:1905048            3.427252e-04                0.9999978          2
7336  GO:0055082            6.193438e-04                0.9999272          6
6709  GO:0048878            8.599730e-04                0.9998733          7
4120  GO:0032879            1.053205e-03                0.9997917         10
3089  GO:0019725            1.440978e-03                0.9997996          6
6965  GO:0051155            1.839005e-03                0.9999690          2
      numInCat                                                        term
11250        3        negative regulation of metalloendopeptidase activity
11386        4            negative regulation of metallopeptidase activity
11249        5                 regulation of metalloendopeptidase activity
5547         4                  regulation of respiratory gaseous exchange
11385        7                     regulation of metallopeptidase activity
7336       248                          intracellular chemical homeostasis
6709       369                                        chemical homeostasis
4120       790                                  regulation of localization
3089       298                                        cellular homeostasis
6965        13 positive regulation of striated muscle cell differentiation
      ontology
11250       BP
11386       BP
11249       BP
5547        BP
11385       BP
7336        BP
6709        BP
4120        BP
3089        BP
6965        BP</code></pre>
</div>
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot them</span></span>
<span id="cb161-2"><a href="#cb161-2" aria-hidden="true" tabindex="-1"></a>bg_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res) <span class="co"># for universe argt in the enrichGO function</span></span>
<span id="cb161-3"><a href="#cb161-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb161-4"><a href="#cb161-4" aria-hidden="true" tabindex="-1"></a>resSIG <span class="ot">&lt;-</span> resDDS <span class="sc">|&gt;</span> <span class="fu">filter</span>(res<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> res<span class="sc">$</span>log2FoldChange <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb161-5"><a href="#cb161-5" aria-hidden="true" tabindex="-1"></a>resSIG_symbol <span class="ot">&lt;-</span> <span class="fu">rownames</span>(resSIG)</span>
<span id="cb161-6"><a href="#cb161-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb161-7"><a href="#cb161-7" aria-hidden="true" tabindex="-1"></a>enrich_go <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(</span>
<span id="cb161-8"><a href="#cb161-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene=</span> resSIG_symbol,</span>
<span id="cb161-9"><a href="#cb161-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">OrgDb =</span> org.Mm.eg.db,</span>
<span id="cb161-10"><a href="#cb161-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">keyType =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb161-11"><a href="#cb161-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">ont =</span> <span class="st">"BP"</span>,</span>
<span id="cb161-12"><a href="#cb161-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">universe =</span> bg_genes,</span>
<span id="cb161-13"><a href="#cb161-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">qvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb161-14"><a href="#cb161-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">readable=</span><span class="cn">TRUE</span></span>
<span id="cb161-15"><a href="#cb161-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb161-16"><a href="#cb161-16" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(enrich_go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="IGF_v1_files/figure-html/unnamed-chunk-37-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>